(GH-1336) Add syntax aware folding provider

glennsarti opened the pull request:
## PR Summary

Fixes #1336 

Adds a syntax aware folding provider, hidden behind a feature flag.

## PR Checklist

Note: Tick the boxes below that apply to this pull request by putting an `x` between the square brackets.
Please mark anything not applicable to this PR `NA`.

Outstanding items
- [x] Outstanding style questions about `.forEach`
- [x] Question about moving `getCoreNodeModule` into `util.js`
- [x] Question about whether to convert FP style to imperative for the `powerShellGrammarPath` function

Required features
- [x] Functions (Fixed in `{ }` support)
- [x] Regions (`#region` and `# region`)
- [x] Here strings
- [x] Comment Blocks  (`# .... \n# ....>`)
- [x] Comment based help (`<# .... #>`)
- [x] Make folding provider optional (via config)

Additional features added
- [x] Hashtable folding (From `{ }` support)
- [x] Array folding (From `( )` support)

PR Review requirements
- [x] PR has a meaningful title
- [x] Summarized changes
- [x] This PR is ready to merge and is not work in progress
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready
- [x] MEGA BONUS POINTS - It has tests!!
- [x] Add logging in the provider/feature
- [x] Converting comments into JSDoc format
- [x] Convert to Literal regex

glennsarti commented:
This is purely just to show the work I've done so far.  It is far from complete (and breaks about 10,000 typescript linting rules :-) )

My current thinking is;
* Use the client side textmate grammar file to tokenize the source
* Once I have that, I can make decisions as to where code begins
* Once I know where code begins and ends I can craft the required `FoldingRange` objects

Note that the syntax folding supersedes any "default" folding regions e.g. `# region` or line break/indentation style.

glennsarti commented:
I have brackets ( `{ }`  and `( )` ) working

![brackets](https://user-images.githubusercontent.com/5819622/41202256-d723890e-6cf8-11e8-8dca-180c5b9d02d3.gif)

It's hard to see but...you'll notice the default folding is there until the extension is loaded, which "magically" removes the folding for the `# region` section.

rkeithhill commented:
> Use the client side textmate grammar file to tokenize the source
Once I have that, I can make decisions as to where code begins

It sounds like you are redoing work the parser has already done.  For example, getting here string folding region info amounts to this:
```
$tok = $err = $null
$ast = [System.Management.Automation.Language.Parser]::ParseFile("$pwd\examples\DebugTest.ps1", [ref]$tok, [ref]$err)
$ast.FindAll({param($ast) ($ast -is [System.Management.Automation.Language.StringConstantExpressionAst]) -and (($ast.StringConstantType -eq [System.Management.Automation.Language.StringConstantType]::DoubleQuotedHereString) -or ($ast.StringConstantType -eq [System.Management.Automation.Language.StringConstantType]::SingleQuotedHereString))}, $true) | % Extent |
Select-Object StartLineNumber, EndLineNumber
```
Which, for my test file outputs:
```
StartLineNumber EndLineNumber
--------------- -------------
              3             9
             34            36
```

We have the ability from the TS extension source to make a request over to the PS Editor Services such that it can `execute PowerShell` code for us including queries on the AST.  This is how the code analysis diagnostics are reported.  This is how help completion works.  This is how "Find PS modules from Gallery" works.  The one area that the AST doesn't help with is comments which includes region support - since that is implemented as "special" comment. Anyway, I had envisioned the folding support would take advantage of the AST for most everything else.  

If you need help with the infrastructure to callback to PSES as well as the PSES C# infrastructure to process the request - let me know.

glennsarti commented:
Hi Keith.  Thanks for taking the time to look at my WIP.  I did consider using the Language Server to surface this information but I decided against it for the following reasons:

1. If the language server crashes, you lose folding.  This is a "breaking change" behaviour of the extension, where folding still works (Indentation based) if the lang. server is unavailable

2. VS Code is the only editor I can see that allows syntax aware folding (Atom and Sublime do not.  Didn't check VIM).  So this feature is specific to VSCode so it seems to belong in the vscode extension, not the generic Editor Services.

3. You can get the region comments by using the tokens generated by the `::ParseFile` method (`$tok`).  But at this point you're replicating things in the lang. server you could do client side.

4. Reduce complexity.  By adding another "thing" to the lang. server. including all the work to add that "thing", and all the plumbing that goes with it, is the increased complexity (and to some degree latency) worth it?.  In my opinion, no.  We can do the same work client side (using the grammar file), with almost exact accuracy as the native PowerShell parser, without the complexity overhead.

Note - By complexity, I'm more referring to the system as a whole, as opposed to complex for a developer to write/understand.

5. (This point is debatable) I feel it makes more sense for the token decisions to come from the textmate grammar than the lang. server. 

`Folding should follow the visual cues of the editor`.  The visual cues come from the colourisation and syntax highlighting, which come the textmate grammar.  These cues do not come from the lang. server.   But ... you could easily argue that the lang. server knows more about the file and should be considered the authority.

Of course there are downsides to going client side:

1. The powershell grammar tokens I'm looking for would become hard-coded.  So if there were changes to the scope names in the grammar file, the folding provider would fail to see them.  This means the client side method is fragile to grammar scope name changes.

Now, is this possible? of course it's possible we'll have changes.  Is this probable? For the tokens I'm looking for, not likely.  Is this testable? Yes, tests could be written to make sure the folding provider continues to function as expected.

rkeithhill commented:
> If the language server crashes, you lose folding.

And intellisense, code analysis, code formatting, code lens, etc.  When the language server crashes, it's a bad day.  I don't think that's a good reason to not utilize PSES for some of this.

>  you could easily argue that the lang. server knows more about the file and should be considered the authority.

Yup. In addition, the tm grammar we've been using has been a steady source of issues/complaints.  I'm not particularly confident in its ability to support PowerShell adequately.   This bug was fixed recently:

![image](https://user-images.githubusercontent.com/5177512/41210029-4dcf5db0-6cec-11e8-8d39-5c77726505ad.png)

For the longest time, Where-Object, Sort-Object, Select-Object would get colorized correctly.  That was recently fixed but there we still get a pretty steady stream of tm grammar issues - see https://github.com/powershell/editorSyntax/issues

I'm probably more than a little jaded about the tm grammar.  :-)  Heck, if I had my druthers we'd use an API like what is being recommended here: https://github.com/Microsoft/vscode/issues/585  Which is essentially how Visual Studio does syntax colorization.  It asks the language service to provide colorization info.

> By adding another "thing" to the lang. server. including all the work to add that "thing", and all the plumbing that goes with it, is the increased complexity (and to some degree latency) worth it?

Depends. If you can provide all the desired folding regions without all the corner cases bugs we seem to so frequently run into with the tm grammar, then no, it's not worth using the AST via the lang server.  OTOH I'd have more confidence in the information we get from AST i.e. I wouldn't worry about corner-cases.  But I'm willing to be proven wrong.  It could be my low opinion of the tm grammar is unwarranted.  :-)

glennsarti commented:
>  Just want to make sure we have a healthy debate about the best approach

Agreed! FWIW I'm one of the authors of the Puppet extension and I have the same debates where to stick features client vs Lang. Server.  Always trade-offs

There is also a middle ground, which is to implement client side until the server side code is complete, and then just switch over.   Could even hide it behind a vscode setting to make it an opt-in feature.  Adding a command to the language server is not trivial, to me anyway.  I am not usually a C# developer so the ramp-up time is not small.  The plumbing to call the server from the client I can handle ok (copy-n-paste to a degree).  If it's decided to go down that route, not client side, I probably won't be able to assist.

Regarding the TM issues... I personally haven't had them, however I too am just a sample size of one :-) 

>  If you can provide all the desired folding regions without all the corner cases bugs we seem to so frequently run into

For the tokens we're interested in in this PR, I don't feel it's too bad, but that's just a gut feel, no data to back it up.

---

So to wrap up.

* The Language Server is probably the final resting place for the folding logic  to be determined, and right now I will not be able to assist with that.

* But I, as a community member, can provide an interim solution using the grammar file for tokenisation, until the more correct solution (lang server) is implemented.  The interim solution can be opt-in which should help with supporting the interim feature.

* Switching over from the client side to server side logic will be transparent to the vscode users.

rkeithhill commented:
Interesting.  So what `FoldingRangeKind` do you use for functions and here strings?

glennsarti commented:
I was going to use 'Region' for everything except comments

rkeithhill commented:
I wonder if we should provide feedback to the VSCode team that we'd like to have a few more "kinds" e.g. function/method, verbatim/here string, or maybe just "Other" as a catchall.

glennsarti commented:
Sure.  I hadn't thought that far ahead.

tylerl0706 commented:
Awesome discussion guys :) I agree that PSES should be, in the end, where the folds come from. Another benefit I don't think was mentioned is that other editors could implement syntax aware folding and those editors can get it for free/minimal effort :)

It would also be good to have all the `using` statements in PowerShell folded as an `imports` type but this is much lower pri IMO.

rjmholt commented:
Hmmmm I definitely see the motivation to make folding all happen client side. Especially since folding should be simple.

But the truth is that trying to do it that way we will inevitably come across things that a TM Grammar is simply not powerful enough to recognise - PowerShell is woefully grammatically complex. The parser is the final arbiter of what text is what token, and if we handle ASTs properly (i.e. cache them, etc), it should all be very cheap too.

@rkeithhill as an aside: I too really want that hook-in highlighting functionality - would be very useful for things like dynamic keyword support and function highlighting.

glennsarti commented:
After happens when the following is in the user settings
``` json
    "powershell.developer.featureFlags": [
        "syntax-folding"
    ],
```

| Before | After |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/5819622/41358094-3916d162-6f5a-11e8-98fc-4fb4ae1c64d1.png) | ![image](https://user-images.githubusercontent.com/5819622/41358104-3e99cdba-6f5a-11e8-9963-3ed6e3c15a6f.png) |

glennsarti commented:
@rkeithhill and @rjmholt PR is done.  Ready for review.

tylerl0706 commented:
@omniomi is the big contributor to [EditorSyntax](https://github.com/powershell/EditorSyntax) which is where the syntax highlighting comes from. 

Nick, if you have some time, can you take a look at this PR and give your thoughts on this since this change depends on EditorSyntax?

omniomi commented:
From a quick look over none of the scopes used for folding in this PR are likely to change as they were all corrected in an older PR. That said, @glennsarti I'll add begin and end scopes to here-strings which I just noticed are missing.

@tylerl0706 I think we're at the point we should create a dependency document for [EditorSyntax](https://github.com/PowerShell/EditorSyntax/) to track the potential fallout from changes. Ie, where it's used, any specific scope dependencies, as well as processes for having dependents update to newer revisions.

tylerl0706 commented on src/features/Folding.ts:
```diff
@@ -0,0 +1,415 @@
+        document: vscode.TextDocument,
+        context: vscode.FoldingContext,
+        token: vscode.CancellationToken,
+    ): Promise<vscode.FoldingRange[]> {
```
nitpick: I noticed some of your functions look like this:

```typescript
public foo(
    doc: bar
    context: baz
) : Boo {
// ...
}
```

and some look like:

```typescript
public foo(doc: bar
           context: baz) : Boo {
// ...
}
```

Is there a reason for the different indenting? Can we align those if not? (I personally prefer the first style 😄)

glennsarti replied:
It's my ruby-isms coming through :-).  Will reformat to first one :-)

glennsarti replied:
Somewhat fixed.  I've reformatted the code I wrote,  haven't done the code I copy-pasta'd from other sources.

tylerl0706 commented on src/features/Folding.ts:
```diff
@@ -0,0 +1,415 @@
+            }
+        });
+
+        return result.reverse();
```
if you need to return the reversed result, maybe use `result.unshift` instead of `result.push`?

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/unshift

glennsarti replied:
Fixed.

tylerl0706 commented on src/features/Folding.ts:
```diff
@@ -0,0 +1,415 @@
+    }
+
+    // Given a zero based offset, return the line number in the document
+    private lineAtOffest(offset: number,
```
looks like a typo: Offest -> Offset

glennsarti replied:
Fixed

tylerl0706 commented on src/features/Folding.ts:
```diff
@@ -0,0 +1,415 @@
+            }
+        });
+
+        // If we exist the token array and we're still processing comment lines, then the
```
did you mean "If we exit"?

glennsarti replied:
Fixed

tylerl0706 commented on src/features/Folding.ts:
```diff
@@ -0,0 +1,415 @@
+                           document: vscode.TextDocument): string {
+        const startPos = document.positionAt(offset);
+        // We don't know how long the line is so just return a really long one.
+        const endPos = startPos.translate(0, 1000);
```
you can get the whole line with:

```typescript
document.lineAt(document.positionAt(offset)).rangeIncludingLineBreak
//or
document.lineAt(document.positionAt(offset)).range // for no LineBreak
```

from that can we accurately get the start and end position?

glennsarti replied:
Fixed

rjmholt commented on src/features/Folding.ts:
```diff
@@ -0,0 +1,415 @@
+            this.foldingProvider);
+    }
+
+    // tslint:disable-next-line:no-empty
```
It looks like these tslint pragmas might be necessary, but I would like to get rid of them if possible. I see there are [recommendations to return `undefined` or `void`](https://stackoverflow.com/questions/31605781/why-do-tslint-and-jslint-report-empty-blocks), but I'm not sold on that either. Thoughts?

rjmholt commented on src/features/Folding.ts:
```diff
@@ -0,0 +1,415 @@
+
+}
+
+export class FoldingFeature implements IFeature {
```
This class should probably be defined at the top of the file.

glennsarti replied:
Was copying the `src/features/DocumentFormatter` style.

rjmholt commented on src/features/Folding.ts:
```diff
@@ -0,0 +1,415 @@
+                                "punctuation.section.group.begin.powershell",
+                                "punctuation.section.group.end.powershell",
+                                MatchType.Region, document)
+            .forEach((x) => { matchedTokens.push(x); });
```
The `x`s used here and below should be renamed more descriptively.

glennsarti replied:
Fixed.

rjmholt commented on src/features/Folding.ts:
```diff
@@ -0,0 +1,415 @@
+                document: vscode.TextDocument) {
+        this.start = start;
+        this.end = end;
+        if (startLine === null) {
```
I was going to suggest `||`, but both suggestions are wrong since `startLine` can be `0`.

Needs to be:
```typescript
this.startLine = startLine !== null ? startLine : document.positionAt(start.startIndex).line;
```

And if there's a worry about the possibility of `undefined` being a possible value of `startLine`, `==` will work for that.

omniomi commented on src/features/Folding.ts:
```diff
@@ -0,0 +1,415 @@
+                                MatchType.Region, document)
+            .forEach((x) => { matchedTokens.push(x); });
+
+        // Find matching brackets   ( -> )
```
I'm inclined to agree and that's as a Canadian where we use the same naming as Australia. (parentheses), {braces}, and [brackets] if we want to follow US convention.

glennsarti replied:
Fixed.  Updated commit message too.  Keeping it US consistent.

glennsarti commented on package.json:
```diff
@@ -5,7 +5,7 @@
   "description": "Develop PowerShell scripts in Visual Studio Code!",
   "engines": {
-    "vscode": "^1.22.0"
+    "vscode": "^1.23.0"
```
Yes.  The Folding API only exists in 1.23.0+

glennsarti replied:
The previous changes didn't separate PRs for it.
https://github.com/PowerShell/vscode-powershell/commit/790e6e2c4f0909a620740e38ef2fc5caf4572fdd

https://github.com/PowerShell/vscode-powershell/commit/5ad413c8f779cc49e6ce9af0dfd43e4eef1bc6cd

https://github.com/PowerShell/vscode-powershell/pull/748

rjmholt replied:
Ok, I'm convinced 😄 

@tylerl0706 Maybe we should consider this a policy for the future, especially with all the feature work going into VSCode — I know .NET Core's latest update was surprisingly tricky for PowerShell.

tylerl0706 replied:
I guess we could. We haven't gotten burned by it yet AFAIK but it doesn't hurt. Next time :)

glennsarti replied:
Next time was last time :-P

glennsarti commented on src/features/Folding.ts:
```diff
@@ -0,0 +1,415 @@
+const tm = getCoreNodeModule("vscode-textmate");
+// Returns a node module installed with VSCode, or null if it fails.
+// https://github.com/Microsoft/vscode/issues/46281
+function getCoreNodeModule(moduleName: string) {
```
Probably, but this is a bit edge-case-y and is only consumed in this file.  Would you still prefer I move it to utils?

glennsarti replied:
@rjmholt I've moved that function as a private method within the feature (to help with loading).  So it's no longer a general function.  Given this, it's it still something I should change for this PR?

glennsarti commented on src/features/Folding.ts:
```diff
@@ -0,0 +1,415 @@
+    // The type of range this is
+    public matchType: MatchType;
+
+    constructor(start: IToken = null,
```
Another possibility ,which in retrospect is better, is to have an empty constructor and use a Builder pattern instead.

glennsarti replied:
Moved to a builder pattern.  Looks nicer too!

glennsarti commented:
Yeah @omniomi , one thing that's missing from this PR is tests.  In theory I should be able to write some simple integration tests (e.g. Given text content <X>, it should return <Y> folding regions).

For better or worse, as this is behind a feature flag, I don't need as good test coverage, yet!.  This _should_ give me enough time to figure out how to download a grammar test fixture, and write appropriate tests.

glennsarti commented on src/features/Folding.ts:
```diff
@@ -0,0 +1,415 @@
+}
+
+// The different types of matched textmate tokens.
+// Mirrors the enum at https://code.visualstudio.com/docs/extensionAPI/vscode-api#FoldingRangeKind
```
I'd rather be consistently wrong, than wrong in inconsistent ways :-).  I'll convert the comments over.

rkeithhill replied:
I'm open to using JSDoc comments on functions - where we actually document said functions.  :-)

glennsarti commented on src/features/Folding.ts:
```diff
@@ -0,0 +1,415 @@
+// Need to reproduce the IToken interface from vscode-textmate due to
+// the odd way it has to be required
+// https://github.com/Microsoft/vscode-textmate/blob/46af9487e1c8fa78aa1f2e2/release/main.d.ts#L161-L165
+export interface IToken {
```
Whoops, that interface is internal only.  Removed the `export`.

glennsarti commented on src/features/Folding.ts:
```diff
@@ -0,0 +1,415 @@
+        // Find blocks of line comments   # comment1\n# comment2\n...
+        this.matchBlockCommentScopeElements(tokens, document).forEach((x) => { matchedTokens.push(x); });
+
+        // Find matching function definitions   <# -> #>
```
Yes.  Updated

glennsarti commented on src/features/Folding.ts:
```diff
@@ -0,0 +1,415 @@
+                                          MatchType.Region, document)
+            .forEach((x) => { matchedTokens.push(x); });
+
+        // Find matching Braces   #region -> #endregion
```
Yes. Updated

glennsarti commented on src/features/Folding.ts:
```diff
@@ -0,0 +1,415 @@
+    private matchGrammarTokens(tokens, document: vscode.TextDocument): IMatchedTokenList {
+        const matchedTokens = [];
+
+        // Find matching Braces   { -> }
```
Fixed

rkeithhill commented on src/features/Folding.ts:
```diff
@@ -0,0 +1,419 @@
+    extensionPath?: string;
+    embeddedLanguages?: {
+        [scopeName: string]: string,
+    }; injectTo?: string[];
```
Should the `injectTo?` field be pushed down onto its own line?

glennsarti replied:
Fixed.

glennsarti commented on src/features/Folding.ts:
```diff
@@ -0,0 +1,415 @@
+        } else {
+            this.startline = startLine;
+        }
+        if (endLine === null) {
```
No longer an issue.

glennsarti commented on src/features/Folding.ts:
```diff
@@ -0,0 +1,415 @@
+    // Given a start and end textmate scope name, find matching grammar tokens
+    // and pair them together.  Uses a simple FIFO stack to take into account
+    // nested regions.
+    private matchScopeElements(tokens,
```
Yeah...on review, because I added IToken, it's trivial to add IToken[].  Fixed.

glennsarti commented on src/features/Folding.ts:
```diff
@@ -0,0 +1,415 @@
+    // Returns the PowerShell grammar parser
+    private grammar() {
+        if (this.powershellGrammar !== undefined) { return this.powershellGrammar; }
+        const registry = new tm.Registry();
```
Yup.  Came to same conclusion too.  I moved the grammar loading from the FoldingProvider, into the FoldingFeature.  This made more sense and should also make testing easier  I made the grammar loading far more defensive, and if it fails, it will not register the folding provider, so at least you'll get the default folding from VSCode.

glennsarti commented on src/features/Folding.ts:
```diff
@@ -0,0 +1,415 @@
+        try {
+            const psGrammars =
+                vscode.extensions.all
+                    .filter((x) => x.packageJSON && x.packageJSON.contributes && x.packageJSON.contributes.grammars)
```
I changed the variable names.  However I didn't convert to imperative, but instead, added code comments as breadcrumbs for future-us.  Would that be enough?

glennsarti commented:
I've addressed most comments.  Outstanding;

* Adding logging
* Outstanding style questions about `.forEach`
* Converting comments into JSDoc format
* Question about moving `getCoreNodeModule` into `util.js`
* Question about whether to convert FP style to imperative for the `powerShellGrammarPath` function

glennsarti commented on src/features/Folding.ts:
```diff
@@ -0,0 +1,415 @@
+                                           document: vscode.TextDocument) {
+        const result = [];
+
+        const emptyLine = new RegExp("^[\\s]+$");
```
Fixed

glennsarti commented:
Moving back to WIP while I fix the remaining items

tylerl0706 commented:
Thank you for your patience, @glennsarti! And for putting up with our comments 😅 A lot of people are going to love this when it gets in 👍

glennsarti commented:
Comments have been converted.  Now waiting of feedback on style questions.

glennsarti commented:
@omniomi I added tests :-)  A single "kitchen sink" integration test.

Test Output;

```
  Features
    Folding Provider
      √ Can detect the PowerShell Grammar
      √ Can detect all of the foldable regions in a document (381ms)
Activating extension 'ms-vscode.PowerShell' failed: command 'PowerShell.ShowLogs' already exists.
Activating extension `ms-vscode.PowerShell` failed:  command 'PowerShell.ShowLogs' already exists
```
Note the `'PowerShell.ShowLogs' already exists.` errors are expected as the Logger cannot be mocked entirely.  This is due to there being no interface, only a concrete implementation.

tylerl0706 commented on src/features/Folding.ts:
```diff
@@ -0,0 +1,602 @@
+
+    /**
+     * Given a start and end textmate scope name, find matching grammar tokens
+     * and pair them together.  Uses a simple FIFO stack to take into account nested regions.
```
> simple FIFO stack

do you mean a queue?

glennsarti replied:
Fixed.  It's a stack.

tylerl0706 commented on src/features/Folding.ts:
```diff
@@ -0,0 +1,602 @@
+        document: vscode.TextDocument,
+    ): string {
+        const startPos: vscode.Position = document.positionAt(offset);
+        const endPos: vscode.Position = document.lineAt(document.positionAt(offset)).range.end;
```
I can't wait for JS to have the pipe operator :D 

https://github.com/tc39/proposal-pipeline-operator

glennsarti replied:
Always be trolling...

glennsarti commented on src/features/Folding.ts:
```diff
@@ -0,0 +1,602 @@
+
+/**
+ * Due to how the vscode-textmate library is required, we need to minimally define a Grammar object, which
+ * can used to tokenize a text document.
```
Fixed

glennsarti commented:
Cross referencing an issue from #1336 

From @mpearon

---

@glennsarti It's working beautifully! A bit of an oddity, though - possibly expected in its current state.
HereStrings are not breaking parent or adjacent folds, but the HereStrings themselves are not folding.

![issue](https://user-images.githubusercontent.com/7157790/41609255-fabc12c4-73af-11e8-9ab5-46e3f034dda4.gif)

glennsarti commented:
@mpearon That still looks like it's using the indentation provider, not the syntax one.  I've added a log entry when the syntax provider is registered in the newest build.  You'll need to grab that and try it.

You should see:
```
2018-6-20 10:34:29 [NORMAL] - Syntax Folding Provider registered
```
In the PowerShell Extension Log - Output window.

glennsarti commented:
And of course, now that @tylerl0706 has fixed Travis, my tests are failing :-)

glennsarti commented:
@tylerl0706 I think tslint is broken or old on Travis

``` bash

> PowerShell@1.7.0 compile /home/travis/build/PowerShell/vscode-powershell
> tsc -v && tsc -p ./ && tslint -p ./ -e test/*
Version 2.9.2
test/features/folding.test.ts:5:26 - error TS2307: Cannot find module '../../src/features/folding'.
5 import * as folding from "../../src/features/folding";
                           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Note that tslint is exlcuding `test/*` and yet the file that failed linting is in `test/...`

tylerl0706 commented:
Travis, you're drunk my friend...

p.s. Thanks for adding tests. We really REALLY lack proper testing in this repo... thanks for not keeping that streak lol

glennsarti commented:
Looks like I'm hitting https://github.com/palantir/tslint/issues/3793

glennsarti commented:
@tylerl0706 @rjmholt I've got the Travis errors sorted and testing is working on AppVeyor.  Just an FYI that extension tests only run on Windows.  Given then current abbreviated test "suite" that's not an issue, but when you start testing per-platform code, you'll have to deal with that.

```
    Folding Provider
      √ Can detect the PowerShell Grammar
      √ Can detect all of the foldable regions in a document (453ms)
Activating extension 'ms-vscode.PowerShell' failed: command 'PowerShell.ShowLogs' already exists.
```

mpearon commented:
@glennsarti So, I must be doing something horribly wrong, because I'm not getting the log entry you mentioned.

-  I'm running Code v1.25.0-insider

- The PowerShell extension is the VSIX you pointed me to, and is registering as ms-vscode.powershell-1.7.0-insiders-616

- The feature flag is thrown

**User Settings:**
`
{
    "extensions.autoUpdate": false,
    "powershell.developer.featureFlags": ["syntax-folding"]
}
`

**PowerShell Extension Logs:**
`
6/20/2018 2:16:27 PM [NORMAL] - Shutting down language client...
6/20/2018 2:16:27 PM [NORMAL] - Terminating PowerShell process...
6/20/2018 2:16:27 PM [NORMAL] - Path specified by 'powerShellExePath' setting - '' - not found, reverting to default PowerShell path.
6/20/2018 2:16:27 PM [NORMAL] - Language server starting --
6/20/2018 2:16:27 PM [NORMAL] -     exe: C:\WINDOWS\System32\WindowsPowerShell\v1.0\powershell.exe
6/20/2018 2:16:27 PM [NORMAL] -     args: C:\Users\USER_NAME\.vscode-insiders\extensions\ms-vscode.powershell-1.7.0-insiders-616\modules\PowerShellEditorServices\Start-EditorServices.ps1 -HostName 'Visual Studio Code Host' -HostProfileId 'Microsoft.VSCode' -HostVersion '1.7.0'-AdditionalModules @('PowerShellEditorServices.VSCode') -BundledModulesPath 'C:\Users\USER_NAME\.vscode-insiders\extensions\ms-vscode.powershell-1.7.0-insiders-616\modules'-EnableConsoleRepl -LanguageServicePipeName LanguageService_29e05d188c4321346c73.pipe -DebugServicePipeName DebugService_29e05d188c4321346c73.pipe -LogLevel 'Normal' -LogPath 'C:\Users\USER_NAME\.vscode-insiders\extensions\ms-vscode.powershell-1.7.0-insiders-616\logs\1529522187-5cf9812d-d898-475e-950a-1d49498817e31529522114627\EditorServices.log' -SessionDetailsPath 'C:\Users\USER_NAME\.vscode-insiders\extensions\ms-vscode.powershell-1.7.0-insiders-616\sessions\PSES-VSCode-17360-681593' -FeatureFlags @('syntax-folding')
6/20/2018 2:16:28 PM [NORMAL] - powershell.exe started, pid: 7472
6/20/2018 2:16:29 PM [NORMAL] - Language server started.
6/20/2018 2:16:29 PM [NORMAL] - {"status":"started","debugServiceTransport":"NamedPipe","languageServiceTransport":"NamedPipe","debugServicePipeName":"DebugService_29e05d188c4321346c73.pipe","languageServicePipeName":"LanguageService_29e05d188c4321346c73.pipe"}
6/20/2018 2:16:29 PM [NORMAL] - Connecting to language service on pipe LanguageService_29e05d188c4321346c73.pipe...
6/20/2018 2:16:29 PM [NORMAL] - Language service connected.
`

tylerl0706 commented:
I still sign off pending we get @mpearon unblocked!

tylerl0706 commented:
@mpearon can you attach all of the logs?

glennsarti commented:
@tylerl0706 Note I logged https://github.com/PowerShell/vscode-powershell/issues/1378 because not all log messages are received.

glennsarti commented:
@mpearon The other way to test is, to not use the packaged extension
Assuming you haven't already;

* Install node
* Clone my repo and checkout my branch
* `npm install`
* Open the source directory
* Run the extension from source

rkeithhill commented:
Do not forget to execute `npm run postinstall` right after `npm install`.

mpearon commented:
@glennsarti and @rkeithhill - I have not tried that method, but I will take a swing at it.  Sorry to be causing headaches!  If it's not worth the troubleshooting (or you just don't feel like dealing with it), feel free to tell me to go away! :)

Verbose Logs attached:
[EditorServices.log](https://github.com/PowerShell/vscode-powershell/files/2123883/EditorServices.log)
[vscode-powershell.log](https://github.com/PowerShell/vscode-powershell/files/2123884/vscode-powershell.log)

tylerl0706 commented:
We would never ask you to go away :) we're here to help!

rjmholt commented:
While I think of it, we should standardise our feature flag naming scheme. I notice `syntax-folding` is Lisp-case, while `PSReadLine` is Pascal-case.

glennsarti commented:
@rjmholt I'd like to see that documented somewhere too.

rkeithhill commented:
Heh, we refer to `syntax-folding` as kebab-case.  I like that visual. :-)

rjmholt commented:
`kebab-case` is probably the better name, more intuitive and universal. Although I admit attachment to using `Lisp-case` just out of hope that Lisp is remembered. Meanwhile, `snake_case` I have never found intuitive. Wikipedia has a nice [rundown on casing](https://en.wikipedia.org/wiki/Letter_case#Special_case_styles), but even there they don't distinguish between `camelCase` and `PascalCase`. It's always the little things that are hard in programming...

tylerl0706 commented:
Casing aside, is there anything outstanding in this PR?

glennsarti commented:
The only things I know of where the 3 points in the PR description.

rjmholt commented:
I'm happy if @rkeithhill is happy 😄

rkeithhill commented:
My only question was whether we should default this to on?  I don't know that many folks will wind up enabling the feature firewall.

tylerl0706 commented:
@rkeithhill I guess this feature can't cause WORSE things to happen lol

I'm a 👍 with enabling by default so long as @glennsarti is ok with adding a config to disable it 😄

glennsarti commented:
@tylerl0706 @rkeithhill I can change it to a setting easy enough. The current setting list is well ordered, but there doesn't appear to be any guidance for contributors about how to choose a correct name. With that in mind, can I please have the EXACT setting name to use from either of you.

rkeithhill commented:
I propose that the setting we use is:

rkeithhill commented:
Although if we think there will not be any other code folding settings, it would be simpler to implement as:
```
"powershell.enableCodeFoldingProvider": true,
```
Then this one setting could be added to ISettings otherwise you need to create a ICodeFoldingSettings interface with one field "enable".  I'm not sure what other settings this feature might have though.  Thoughts?

tylerl0706 commented:
that's a good point. @glennsarti do you see any sort of extensibility/configurable nature in the CodeFoldingProvider? Or some configuration that could be applied to the future?

I could think that maybe (in the future) we could allow the user to specify a list of things they'd want to fold... but maybe that's not really something someone would want...

glennsarti commented:
@tylerl0706 My gut feeling is no. There are no future config settings; just the binary On/Off.  However once the code folding goes from client to language server, all bets are off as you have much more control there.  Given unknown future and the small amount of code difference between the two options, I'd go the safe route with `"powershell.codeFolding.enable": true,`

tylerl0706 commented:
Sounds good to me 👍 on:
`"powershell.codeFolding.enable": true,`

When it gets moved to the language server side, that can be behind a feature flag initially since it could make code folding worse.

glennsarti commented:
@tylerl0706 @rkeithhill  @rjmholt Done

With `"powershell.codeFolding.enable": false,` in the config you get the following in the logs;
```
2018-6-28 13:01:16 [NORMAL] - Visual Studio Code v1.24.1 64-bit
2018-6-28 13:01:16 [NORMAL] - PowerShell Extension v1.7.0
2018-6-28 13:01:16 [NORMAL] - Operating System: Windows 64-bit
2018-6-28 13:01:16 [NORMAL] - Path specified by 'powerShellExePath' setting - '' - not found, reverting to default PowerShell path.
```

With `"powershell.codeFolding.enable": true,` or not set at all, in the config you get the following in the logs;
```
2018-6-28 13:00:00 [NORMAL] - Visual Studio Code v1.24.1 64-bit
2018-6-28 13:00:00 [NORMAL] - PowerShell Extension v1.7.0
2018-6-28 13:00:00 [NORMAL] - Operating System: Windows 64-bit
2018-6-28 13:00:00 [NORMAL] - Syntax Folding Provider registered    <----- Syntax folder
2018-6-28 13:00:00 [NORMAL] - Path specified by 'powerShellExePath' setting - '' - not found, reverting to default PowerShell path.
```

And of course you get syntax vs indentation based folding as you'd expect given the log files

tylerl0706 commented:
@glennsarti thank you for putting up with us lol this is amazing and I'm really excited to get this in :tada:

rjmholt commented on src/features/Folding.ts:
```diff
@@ -0,0 +1,601 @@
+                        return ext.packageJSON && ext.packageJSON.contributes && ext.packageJSON.contributes.grammars;
+                    })
+                    // For each extension, extract the list of grammar languages
+                    .reduce((grammarList: IExtensionGrammar[], ext) =>
```
The `reduce` here still seems a bit complicated to me, given the spread operators and the internal `map` it uses. It's annoying because I love a good fold, but I think for maintainability's sake we should change it. Maybe to something like:

```typescript
// Go through all the extension packages and search for PowerShell grammars,
// returning the path to the first we find
foreach (const ext of vscode.extensions.all) {
    if (!(ext.packageJSON && ext.packageJSON.contributes && ext.packageJSON.contributes.grammars)) {
        continue;
    }

    foreach (const grammar: IGrammarExtension of (ext.packageJSON as IExtensionPackage).contributes.grammars) {
        if (grammar.language !== "powershell") {
            continue;
        }

        return path.join(ext.extensionPath, grammar.path);
    }
}

return undefined;
```

This way, we return when the first PowerShell grammar is found and there's no array copying with each accumulator pass.

(If I've totally misunderstood what this method does, just ignore me 😄)

glennsarti commented on package.json:
```diff
@@ -467,6 +467,11 @@
+        "powershell.codeFolding.enable": {
+          "type": "boolean",
+          "default": true,
+          "description": "Enables syntax based code folding.  When disabled the default indentation based code folding."
```
Fixed

glennsarti commented on src/features/Folding.ts:
```diff
@@ -0,0 +1,601 @@
+        const grammarPath = this.powerShellGrammarPath();
+        if (grammarPath == null) {
+            return undefined;
+        } else {
```
Fixed

glennsarti commented on src/features/Folding.ts:
```diff
@@ -0,0 +1,601 @@
+                        startLine = lineNum;
+                        nextLine = lineNum + 1;
+                    } else {
+                        if (lineNum === nextLine) {
```
Yes, and I managed to compress it even further.

glennsarti commented:
@tylerl0706 @rjmholt Addressed all comments.  CI is green...

rkeithhill commented on src/features/Folding.ts:
```diff
@@ -0,0 +1,588 @@
+import * as path from "path";
```
What's the story on source file copyright notices?  Every other .ts file seems to have this copyright notice on the first line:
```
/*---------------------------------------------------------
 * Copyright (C) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------*/

```
Even those contributed by community members.

glennsarti replied:
❓ Contributing guide/s makes no mention.  I've already signed the CLA.

glennsarti replied:
I, as a non-Microsoft employee, shouldn't add that, as I don't represent them.

rkeithhill replied:
I don't mention this as a blocker but an inquiry as to what the policy is.  Back in January I went through this code-base and cleaned up source to pass tslint as part of effort to achieve some form of engineering compliance  (#1086).  I'm curious if that "compliance" has anything to say about copyright notices.

SeeminglyScience replied:
> I, as a non-Microsoft employee, shouldn't add that, as I don't represent them.

FWIW the copyright is their's either way, and ultimately the one representing Microsoft is whomever publishes the release.

I view it more as "informing of a copyright" instead of "declaring copyright".

That said, I'm not sure what the point of it is - at least in an MIT licensed project.  Maybe @tylerl0706 or @rjmholt could check if there is a policy on this?  If there isn't, my vote would be to remove it or at least omit it going forward.

(This is just a general note, definitely shouldn't hold up the PR)

tylerl0706 replied:
The copyright header is definitely required. If you don't want to add it, I can do a quick amend.

tylerl0706 replied:
This is one of those "I don't make the rules, I just (have to) follow them

rjmholt replied:
We should definitely add a note in the contribution guide...

tylerl0706 replied:
added the headers

rjmholt replied:
Btw, I looked at our contriibution guide and was about to add a note but wasn't sure of where to add it or how to phrase it. I also notice that PowerShell doesn't seem to have a note about it either, only in their PR checklist.

tylerl0706 replied:
We could add it to the PR checklist :)

glennsarti replied:
Offtopic - @SeeminglyScience - Ceeding copyright to Microsoft is fine.  It was more the "All rights reserved".  IANAL, but I didn't feel I was in position to represent to Microsoft to say what rights other people had. Again this all moot (Heh, I used legal term!) due to me signing the CLA.

glennsarti replied:
@tylerl0706 @rjmholt Raised #1394 to add automation to enforce header rule.

tylerl0706 commented:
Merging this in tomorrow morning :) unless people have further comments

tylerl0706 commented:
It's "Squash" not "Smash" psh.... Hulk, please git good.

This is merged!! :smile:

glennsarti commented:
![No Squash](https://i.imgflip.com/2db3g6.jpg)
